textpipe/doc.py

```diff
@@ -413,3 +414,78 @@ def similarity(self, other_doc, metric='jaccard', hash_method='minhash'):
         else:
             raise NotImplementedError(f'Metric/hash method combination {metric}'
                                       f'/{hash_method} is not implemented as similarity metric')
+
+    @property
+    def word_vectors(self):
+        """
+        Returns word embeddings for the words in the document.
+        The used spacy models don't have "true" word vectors
+        but only context-sensitive tensors that are within the document.
+
+        Returns:
+        A dictionary mapping words from the document to a dict with the
+        corresponding values of the following variables:
+
+        has vector: Does the token have a vector representation?
+        vector norm: The L2 norm of the token's vector (the square root of the
+                    sum of the values squared)
+        OOV: Out-of-vocabulary (This variable always gets the value True since
+                                there are no vectors included in the model)
+        vector: The vector representation of the word
+
+        >>> doc = Doc('Test sentence')
+        >>> doc.word_vectors['Test']['is_oov']
+        True
+        >>> doc.word_vectors['Test']['vector'].shape[-1]
+        384
+        >>> doc.word_vectors['Test']['vector_norm'] == doc.word_vectors['sentence']['vector_norm']
+        False
+        """
+        return {token.text: {'has_vector': token.has_vector,
+                             'vector_norm': token.vector_norm,
+                             'is_oov': token.is_oov,
+                             'vector': token.vector}
+                for token in self._spacy_doc}
+
+    @property
+    def doc_vector(self):
+        """
+        Returns document embeddings based on the words in the document.
+
+        >>> import numpy
+        >>> numpy.array_equiv(Doc('a b').doc_vector, Doc('a b').doc_vector)
+        True
+        >>> numpy.array_equiv(Doc('a b').doc_vector, Doc('a a b').doc_vector)
+        False
+        """
+        return self.aggregate_word_vectors()
+
+    def aggregate_word_vectors(self, aggregation='mean', normalize=False, exclude_oov=False):
+        """
+        Returns document embeddings based on the words in the document.
```

Comment:
A more general comment: in our use-case, we want to compute multiple aggregations of a doc, to be precise:

```
        vector_list = [token.vector for token in doc_vectors]
        doc_vectors_mean = doc_vectors.vector.tolist()  # the same as np.mean(vector_list, axis=0)
        doc_vectors_sum = np.sum(vector_list, axis=0).tolist()
        doc_vectors_var = np.var(vector_list, axis=0).tolist()
        doc_vectors_l2norm = doc_vectors.vector_norm

        return {'word2vec_average': doc_vectors_mean,
                'word2vec_sum': doc_vectors_sum,
                'word2vec_variation': doc_vectors_var,
                'word2vec_l2norm': doc_vectors_l2norm}
```

In the current implementation, that would mean redoing the aggregation for each --- something to consider?

Reply:
I don't think generating the vector_list is the slowest thing here. It's the mean, sum and var operations. We could benchmarkt this, but I would opt too leave as is. We could at the var operation, if you are actually using that.